File: server/src/main/java/com/mandilas/market/service/OrderService.java
package com.mandilas.market.service;

import com.mandilas.market.model.Order;
import com.mandilas.market.model.OrderItem;
import com.mandilas.market.repository.OrderRepository;

import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class OrderService {

    private final OrderRepository orderRepository;

    public OrderService(
            OrderRepository orderRepository
    ) {

        this.orderRepository =
                orderRepository;
    }


    /*
     * =========================================================
     * CREATE A NEW ORDER
     * =========================================================
     */
    public Order createOrder(
            Order order
    ) {

        if (
                order == null
        ) {

            throw new RuntimeException(
                    "Order is required"
            );
        }


        if (
                order.getItems() == null ||
                order.getItems().isEmpty()
        ) {

            throw new RuntimeException(
                    "Order must contain at least one product"
            );
        }


        /*
         * Make sure every OrderItem points back
         * to this Order.
         */
        for (
                OrderItem item :
                order.getItems()
        ) {

            if (item != null) {

                item.setOrder(order);
            }
        }


        /*
         * Calculate each item's total.
         */
        for (
                OrderItem item :
                order.getItems()
        ) {

            if (item == null) {
                continue;
            }

            double itemTotal =
                    item.getPrice()
                            * item.getQuantity();

            item.setTotal(
                    itemTotal
            );
        }


        /*
         * Calculate subtotal.
         */
        double subtotal =
                order.getItems()
                        .stream()
                        .filter(
                                item -> item != null
                        )
                        .mapToDouble(
                                OrderItem::getTotal
                        )
                        .sum();

        order.setSubtotal(
                subtotal
        );


        /*
         * Calculate final order total.
         */
        double total =
                order.getSubtotal()
                        + order.getDeliveryFee();

        order.setTotal(
                total
        );


        /*
         * Default payment status.
         */
        if (
                order.getPaymentStatus() == null ||
                order.getPaymentStatus().isBlank()
        ) {

            order.setPaymentStatus(
                    "Pending"
            );
        }


        /*
         * Default order status.
         */
        if (
                order.getOrderStatus() == null ||
                order.getOrderStatus().isBlank()
        ) {

            order.setOrderStatus(
                    "Pending"
            );
        }


        return orderRepository.save(
                order
        );
    }


    /*
     * =========================================================
     * GET ALL ORDERS
     * =========================================================
     */
    public List<Order> getAllOrders() {

        return orderRepository.findAll();
    }


    /*
     * =========================================================
     * GET ONE ORDER
     * =========================================================
     */
    public Order getOrderById(
            Long id
    ) {

        if (id == null) {

            throw new RuntimeException(
                    "Order ID is required"
            );
        }

        return orderRepository
                .findById(id)
                .orElseThrow(
                        () ->
                                new RuntimeException(
                                        "Order not found"
                                )
                );
    }


    /*
     * =========================================================
     * GET CUSTOMER ORDERS
     * =========================================================
     */
    public List<Order> getOrdersByEmail(
            String email
    ) {

        if (
                email == null ||
                email.trim().isEmpty()
        ) {

            throw new RuntimeException(
                    "Customer email is required"
            );
        }

        return orderRepository.findByEmail(
                email
                        .trim()
                        .toLowerCase()
        );
    }


    /*
     * =========================================================
     * GET SELLER ORDERS
     * =========================================================
     *
     * NO JWT.
     * NO Authentication.
     *
     * Seller identity is supplied by the controller.
     *
     * Orders are matched through:
     *
     * OrderItem.sellerEmail
     *
     * =========================================================
     */
    public List<Order> getOrdersBySellerEmail(
            String sellerEmail
    ) {

        if (
                sellerEmail == null ||
                sellerEmail.trim().isEmpty()
        ) {

            throw new RuntimeException(
                    "Seller email is required"
            );
        }


        return orderRepository
                .findDistinctByItemsSellerEmail(
                        sellerEmail
                                .trim()
                                .toLowerCase()
                );
    }


    /*
     * =========================================================
     * GET ORDERS BY ORDER STATUS
     * =========================================================
     */
    public List<Order> getOrdersByStatus(
            String orderStatus
    ) {

        if (
                orderStatus == null ||
                orderStatus.trim().isEmpty()
        ) {

            throw new RuntimeException(
                    "Order status is required"
            );
        }

        return orderRepository
                .findByOrderStatus(
                        orderStatus
                                .trim()
        );
    }


    /*
     * =========================================================
     * GET ORDERS BY PAYMENT STATUS
     * =========================================================
     */
    public List<Order> getOrdersByPaymentStatus(
            String paymentStatus
    ) {

        if (
                paymentStatus == null ||
                paymentStatus.trim().isEmpty()
        ) {

            throw new RuntimeException(
                    "Payment status is required"
            );
        }

        return orderRepository
                .findByPaymentStatus(
                        paymentStatus
                                .trim()
                );
    }


    /*
     * =========================================================
     * UPDATE ORDER STATUS
     * =========================================================
     */
    public Order updateOrderStatus(
            Long id,
            String orderStatus
    ) {

        if (
                orderStatus == null ||
                orderStatus.trim().isEmpty()
        ) {

            throw new RuntimeException(
                    "Order status is required"
            );
        }


        Order order =
                getOrderById(id);


        order.setOrderStatus(
                orderStatus.trim()
        );


        return orderRepository.save(
                order
        );
    }


    /*
     * =========================================================
     * UPDATE PAYMENT STATUS
     * =========================================================
     */
    public Order updatePaymentStatus(
            Long id,
            String paymentStatus
    ) {

        if (
                paymentStatus == null ||
                paymentStatus.trim().isEmpty()
        ) {

            throw new RuntimeException(
                    "Payment status is required"
            );
        }


        Order order =
                getOrderById(id);


        order.setPaymentStatus(
                paymentStatus.trim()
        );


        return orderRepository.save(
                order
        );
    }


    /*
     * =========================================================
     * DELETE ORDER
     * =========================================================
     */
    public void deleteOrder(
            Long id
    ) {

        if (
                id == null
        ) {

            throw new RuntimeException(
                    "Order ID is required"
            );
        }


        if (
                !orderRepository.existsById(id)
        ) {

            throw new RuntimeException(
                    "Order not found"
            );
        }


        orderRepository.deleteById(
                id
        );
    }
}

File: client/src/services/orderService.js
import API_URL from "./api";


/*
 * =========================================================
 * GET SIMPLE USER SESSION
 * =========================================================
 *
 * No JWT.
 *
 * Seller/customer identity comes from:
 *
 * mandilas-user
 *
 * =========================================================
 */
function getSavedUser() {

  const savedUser =
    localStorage.getItem(
      "mandilas-user"
    );

  if (!savedUser) {
    return null;
  }

  try {

    return JSON.parse(
      savedUser
    );

  } catch {

    return null;
  }
}


/*
 * =========================================================
 * CREATE ORDER
 * =========================================================
 */
export async function createOrder(order) {

  const response =
    await fetch(
      `${API_URL}/orders`,
      {
        method: "POST",

        headers: {
          "Content-Type":
            "application/json",
        },

        body:
          JSON.stringify(order),
      }
    );


  if (!response.ok) {

    const errorText =
      await response.text();

    throw new Error(
      errorText ||
        "Failed to create order"
    );
  }

  return response.json();
}


/*
 * =========================================================
 * GET ALL ORDERS
 * =========================================================
 */
export async function getAllOrders() {

  const response =
    await fetch(
      `${API_URL}/orders`,
      {
        method: "GET",
      }
    );


  if (!response.ok) {

    const errorText =
      await response.text();

    throw new Error(
      errorText ||
        "Failed to fetch orders"
    );
  }

  return response.json();
}


/*
 * =========================================================
 * GET ONE ORDER
 * =========================================================
 */
export async function getOrderById(id) {

  const response =
    await fetch(
      `${API_URL}/orders/${id}`,
      {
        method: "GET",
      }
    );


  if (!response.ok) {

    const errorText =
      await response.text();

    throw new Error(
      errorText ||
        "Failed to fetch order"
    );
  }

  return response.json();
}


/*
 * =========================================================
 * GET CUSTOMER ORDERS
 * =========================================================
 */
export async function getOrdersByEmail(
  email
) {

  if (
    !email ||
    !email.trim()
  ) {

    throw new Error(
      "Customer email is required"
    );
  }


  const response =
    await fetch(
      `${API_URL}/orders/customer/${encodeURIComponent(
        email.trim()
      )}`,
      {
        method: "GET",
      }
    );


  if (!response.ok) {

    const errorText =
      await response.text();

    throw new Error(
      errorText ||
        "Failed to fetch customer orders"
    );
  }

  return response.json();
}


/*
 * =========================================================
 * GET SELLER ORDERS
 * =========================================================
 *
 * NO JWT.
 * NO Authorization header.
 *
 * Seller email comes from:
 *
 * mandilas-user
 *
 * Request:
 *
 * GET /api/orders/seller?email=seller@email.com
 *
 * =========================================================
 */
export async function getOrdersBySellerEmail() {

  const user =
    getSavedUser();


  const sellerEmail =
    user?.email || "";


  if (!sellerEmail) {

    throw new Error(
      "Please login as a seller."
    );
  }


  const response =
    await fetch(
      `${API_URL}/orders/seller?email=${encodeURIComponent(
        sellerEmail.trim()
      )}`,
      {
        method: "GET",
      }
    );


  if (!response.ok) {

    const errorText =
      await response.text();

    throw new Error(
      errorText ||
        "Failed to fetch seller orders"
    );
  }

  return response.json();
}


/*
 * =========================================================
 * GET ORDERS BY ORDER STATUS
 * =========================================================
 */
export async function getOrdersByStatus(
  status
) {

  if (
    !status ||
    !status.trim()
  ) {

    throw new Error(
      "Order status is required"
    );
  }


  const response =
    await fetch(
      `${API_URL}/orders/status/${encodeURIComponent(
        status.trim()
      )}`,
      {
        method: "GET",
      }
    );


  if (!response.ok) {

    const errorText =
      await response.text();

    throw new Error(
      errorText ||
        "Failed to fetch orders by status"
    );
  }

  return response.json();
}


/*
 * =========================================================
 * GET ORDERS BY PAYMENT STATUS
 * =========================================================
 */
export async function getOrdersByPaymentStatus(
  status
) {

  if (
    !status ||
    !status.trim()
  ) {

    throw new Error(
      "Payment status is required"
    );
  }


  const response =
    await fetch(
      `${API_URL}/orders/payment-status/${encodeURIComponent(
        status.trim()
      )}`,
      {
        method: "GET",
      }
    );


  if (!response.ok) {

    const errorText =
      await response.text();

    throw new Error(
      errorText ||
        "Failed to fetch orders by payment status"
    );
  }

  return response.json();
}


/*
 * =========================================================
 * UPDATE ORDER STATUS
 * =========================================================
 */
export async function updateOrderStatus(
  id,
  status
) {

  if (
    !status ||
    !status.trim()
  ) {

    throw new Error(
      "Order status is required"
    );
  }


  const response =
    await fetch(
      `${API_URL}/orders/${id}/status?status=${encodeURIComponent(
        status.trim()
      )}`,
      {
        method: "PUT",
      }
    );


  if (!response.ok) {

    const errorText =
      await response.text();

    throw new Error(
      errorText ||
        "Failed to update order status"
    );
  }

  return response.json();
}


/*
 * =========================================================
 * UPDATE PAYMENT STATUS
 * =========================================================
 */
export async function updatePaymentStatus(
  id,
  status
) {

  if (
    !status ||
    !status.trim()
  ) {

    throw new Error(
      "Payment status is required"
    );
  }


  const response =
    await fetch(
      `${API_URL}/orders/${id}/payment-status?status=${encodeURIComponent(
        status.trim()
      )}`,
      {
        method: "PUT",
      }
    );


  if (!response.ok) {

    const errorText =
      await response.text();

    throw new Error(
      errorText ||
        "Failed to update payment status"
    );
  }

  return response.json();
}


/*
 * =========================================================
 * DELETE ORDER
 * =========================================================
 */
export async function deleteOrder(id) {

  const response =
    await fetch(
      `${API_URL}/orders/${id}`,
      {
        method: "DELETE",
      }
    );


  if (!response.ok) {

    const errorText =
      await response.text();

    throw new Error(
      errorText ||
        "Failed to delete order"
    );
  }


  /*
   * DELETE may return 204 No Content.
   */
  if (
    response.status === 204
  ) {

    return true;
  }


  return response.json();
}